src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/OOPTagHelperResolver.cs

```diff
@@ -105,14 +109,58 @@ protected virtual async Task<TagHelperResolutionResult> ResolveTagHelpersOutOfPr
                 return null;
             }
 
+            if (!_resultCache.TryGetId(projectSnapshot.FilePath, out var lastResultId))
+            {
+                lastResultId = -1;
+            }
+
             var projectHandle = new ProjectSnapshotHandle(projectSnapshot.FilePath, projectSnapshot.Configuration, projectSnapshot.RootNamespace);
-            var result = await remoteClient.TryInvokeAsync<IRemoteTagHelperProviderService, TagHelperResolutionResult>(
+            var result = await remoteClient.TryInvokeAsync<IRemoteTagHelperProviderService, TagHelperDeltaResult>(
                 workspaceProject.Solution,
-                (service, solutionInfo, innerCancellationToken) => service.GetTagHelpersAsync(solutionInfo, projectHandle, factory?.GetType().AssemblyQualifiedName, innerCancellationToken),
+                (service, solutionInfo, innerCancellationToken) => service.GetTagHelpersDeltaAsync(solutionInfo, projectHandle, factory?.GetType().AssemblyQualifiedName, lastResultId, innerCancellationToken),
                 cancellationToken
             );
 
-            return result.HasValue ? result.Value : null;
+            if (!result.HasValue)
+            {
+                return null;
+            }
+
+            var tagHelpers = ConsumeTagHelpersFromDelta(projectSnapshot.FilePath, lastResultId, result.Value);
+
+            var resolutionResult = new TagHelperResolutionResult(tagHelpers, diagnostics: Array.Empty<RazorDiagnostic>());
+            return resolutionResult;
+        }
+
+        // Protected virtual for testing
+        protected virtual IReadOnlyList<TagHelperDescriptor> ConsumeTagHelpersFromDelta(string projectFilePath, int lastResultId, TagHelperDeltaResult deltaResult)
```

Comment:
I definitely struggled with naming this. I initially chose consume because I'm trying to represent the "producing" + "caching" of what's going on behind the scenes. That being said I totally agree it aint ideal. Produce aint bad